File: python_version/insertion_sort.py
#!/usr/bin/env python3


def insertion_sort(arr):
    '''sort the arr by insertion sort'''
    for i in range(1, len(arr)):
        gap = i
        temp = arr[i]
        while gap >= 1 and arr[gap - 1] > temp:
            arr[gap] = arr[gap - 1]
            gap -= 1
        arr[gap] = temp


arr = [2, 4, 1, 5, 12, 0, 23]
print(arr)
insertion_sort(arr)
print(arr)

File: cpp_version/insertion_sort.cpp
//Worst Case Complexity in time: O(n^2)
//Average Case Complexity in time: O(n^2)
//Best Case Complexity in time: O(n)
//Space Complexity: O(1)
//The insertion sort is used when:
//the array is has a small number of elements
//there are only a few elements left to be sorted
#include <iostream>
#include <vector>
#include <algorithm>

#include "header.h"

using namespace std;

static Count res;

Count InsertionSort(vector<int> &vec)
{
    int len = vec.size();
    for (int i = 1; i < len; i++)
    {
        //空位下标为i, 空位值为temp
        int gap = i;
        int temp = vec[i];
        while (gap >= 1 && vec[gap - 1] > temp)
        {
            res.compare++;
            //赋值给右边
            vec[gap] = vec[gap - 1];
            --gap;
            res.move++;
        }
        vec[gap] = temp;
        res.move++;
    }
    return res;
}
